libvips/foreign/dzsave.c

```diff
@@ -866,61 +529,37 @@ write_image( VipsForeignSaveDz *dz,
 	 * off. Most people don't want metadata on every tile.
 	 */
 	vips_image_set_int( t, "hide-progress", 1 );
-	if( vips_image_write_to_buffer( t, format, &buf, &len,
-		"strip", !dz->no_strip,
-		NULL ) ) {
-		VIPS_UNREF( t );
-		return( -1 );
-	}
-	VIPS_UNREF( t );
 
-	/* gsf doesn't like more than one write active at once.
-	 */
-	g_mutex_lock( vips__global_lock );
-
-	if( !gsf_output_write( out, len, buf ) ) {
-		gsf_output_close( out );
-		g_mutex_unlock( vips__global_lock );
-		g_free( buf );
-
-		if( gsf_output_error( out ) )
-			vips_error( class->nickname,
-				"%s", gsf_output_error( out )->message );
-
-		return( -1 );
-	}
-
-	dz->bytes_written += len;
-
-	gsf_output_close( out );
+#ifdef HAVE_ZIP
+	if( iszip( dz->container ) ) {
+		void *buf;
+		size_t len;
 
-#ifndef HAVE_GSF_ZIP64
-	if( iszip( dz->container ) ) { 
-		/* Leave 3 entry headroom for blank.png and metadata files.
-		 */
-		if( dz->tree->file_count + 3 >= (unsigned int) USHRT_MAX ) {
-			g_mutex_unlock( vips__global_lock );
-
-			vips_error( class->nickname,
-				"%s", _( "too many files in zip" ) );
+		if( vips_image_write_to_buffer( t, dz->suffix, &buf, &len,
```

Comment:
This needs to be revisited, `blank.png` should always be written as PNG when `dz->suffix` is `".jpeg"`. The same applies to associated images, they should probably always be written as JPEG (i.e. we cannot assume that all image types are the same).